<a href="https://colab.research.google.com/github/Victoriambaka/IP-WEEK-3/blob/main/Moringa_Data_Science_Prep_W3_Independent_Project_2022_02_Victoria_Mbaka_Python_Notebook_resubmission_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# INDEPENDENT PROJECT WEEK 3 RESUBMISSION
## MTN Cote d’Ivoire/Ivory Coast requires to upgrade their infrastructure for its clients. A data scientist is required to process their data and answer the following research questions and more.

* Which ones were the most used city for the three days?
* Which cities were the most used during business and home hours?
* Most used city for the three days? 


Datasets:
The telecom data provided for this project is only a sample ( i.e. for only three days).  The data files that you will need for this project will be as follows:

* cells_geo_description.xlsx [Link] (Links to an external site.)
* cells_geo.csv [Link] (Links to an external site.)
* CDR_description.xlsx [Link] (Links to an external site.)
* CDR 20120507 [http://bit.ly/TelecomDataset1] (Links to an external site.)
* CDR 20120508 [http://bit.ly/TelecomDataset2] (Links to an external site.)
* CDR 20120509 [http://bit.ly/TelecomDataset3]



# Importing needed libraries for the project

In [37]:
# Importing pandas and numpy libraries
import pandas as pd
import numpy as np

# Loading the datasets 


In [38]:
# We will load the datasets and convert the to dataframes.
# Loading cells_geo dataset as cgeo
cgeo = pd.read_csv("cells_geo.csv",sep =';',index_col=0)
cgeo.head()

VILLES      STATUS LOCALISATION  ... AREA     CELL_ID   SITE_CODE
0   ADJAME  In Service      ABIDJAN  ...  NaN  fbb2e75af3  0bb492ef17
1   ADJAME  In Service      ABIDJAN  ...  NaN  1502501d78  0bb492ef17
2   ADJAME  In Service      ABIDJAN  ...  NaN  0c2e182a0c  0bb492ef17
3  ASSINIE  In Service    INTERIEUR  ...  NaN  69bdfc69fd  8a1e6ce31f
4  ASSINIE  In Service    INTERIEUR  ...  NaN  7a1029d2ab  8a1e6ce31f

[5 rows x 11 columns]

In [39]:
# Loading cells_geo_description dataset as geographic
geographic = pd.read_excel("cells_geo_description.xlsx")
geographic

Column name                                        Description  Format
0         VILLES                                               City  String
1         STATUS                                  In Service or not  String
2   LOCALISATION                                  in ABIDJAN or not  String
3     DECOUPZONE                                  Geographical Zone  String
4       ZONENAME                                       Name of Zone  String
5      LONGITUDE                                          Longitude   Float
6       LATITUDE                                           Latitude   Float
7         REGION                                             Region  String
8           AREA                                               Area  String
9        CELL_ID                                     ID of the cell  String
10     SITE_CODE  Site (there are several cells per site, severa...  String

In [40]:
# Loading CDR_description as CDR
CDR = pd.read_excel("CDR_description.xlsx")
CDR

Column name                                        Description   Format
0          PRODUCT                                       Voice or SMS   String
1            VALUE                                      Billing price  Integer
2        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
3     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
4  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String
5  DW_B_NUMBER_INT        Anonymized phone number of the counterparty   String
6        COUNTRY_A                  Country of party A (useless here)   String
7        COUNTRY_B                  Country of party B (useless here)   String
8          CELL_ID                                     ID of the cell   String
9          SITE_ID                                     ID of the SITE   String

In [41]:
# Loading Telcom dataset as Tel_1
Tel_1 = pd.read_csv("Telcom_dataset.csv")
Tel_1.head()

PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0   Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv

[5 rows x 10 columns]

In [42]:
# Loading Telcom dataset 2 as Tel_2
Tel_2 = pd.read_csv("Telcom_dataset2.csv")
Tel_2.head()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0     sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1   Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2     sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3   Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4     sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO

[5 rows x 10 columns]

In [43]:
# Loading Telcom dataset 3 as Tel_3
Tel_3 = pd.read_csv("Telcom_dataset3.csv")
Tel_3.head()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SIET_ID
0   Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1     sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2   Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3     sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA

[5 rows x 10 columns]

# Data Understanding

In data understanding, we are finding out specific information from our datasets 

In [44]:
# understanding data from Cells_geo
# Finding out data types in the dataset cgeo.
print(cgeo.info())


<class 'pandas.core.frame.DataFrame'>
Int64Index: 3974 entries, 0 to 3973
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   VILLES        3974 non-null   object 
 1   STATUS        3907 non-null   object 
 2   LOCALISATION  3974 non-null   object 
 3   DECOUPZONE    3974 non-null   object 
 4   ZONENAME      3968 non-null   object 
 5   LONGITUDE     3974 non-null   float64
 6   LATITUDE      3974 non-null   float64
 7   REGION        3974 non-null   object 
 8   AREA          3951 non-null   object 
 9   CELL_ID       3974 non-null   object 
 10  SITE_CODE     3974 non-null   object 
dtypes: float64(2), object(9)
memory usage: 372.6+ KB
None


* cells_geo dataset has no null values
* all values are categorical except two, which are floats

In [45]:
# Finding out sum of null values from cgeo
print(cgeo.isnull().sum())

VILLES           0
STATUS          67
LOCALISATION     0
DECOUPZONE       0
ZONENAME         6
LONGITUDE        0
LATITUDE         0
REGION           0
AREA            23
CELL_ID          0
SITE_CODE        0
dtype: int64


* cgeo has 67 null values in Status column, 6 in Zonename column and 23 in area column.

In [46]:
# Finding out data types from Telcom dataset(Tel_1)
print(Tel_1.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PRODUTC          5001 non-null   object
 1   VALUE            5001 non-null   int64 
 2   DATETIME         5001 non-null   object
 3   CELL_ON_SITE     5001 non-null   object
 4   DW_A_NUMBER_INT  5001 non-null   object
 5   DW_B_NUMBER_INT  5001 non-null   object
 6   COUNTRY_A        5001 non-null   object
 7   COUNTRY_B        5001 non-null   object
 8   CELL_ID          5001 non-null   object
 9   SITE_ID          4285 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB
None


* Tel_1 has no missing values except for 716 entries in Site_id column

In [47]:
# Finding out data types inTelcom dataset 2 (Tel_2)
print(Tel_2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   PRODUCT       5001 non-null   object
 1   VALUE         5001 non-null   int64 
 2   DATE_TIME     5001 non-null   object
 3   CELL_ON_SITE  5001 non-null   object
 4   DW_A_NUMBER   5001 non-null   object
 5   DW_B_NUMBER   5001 non-null   object
 6   COUNTRY_A     5001 non-null   object
 7   COUNTRY_B     5001 non-null   object
 8   CELL_ID       5001 non-null   object
 9   SITE_ID       4325 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB
None


* Tel_2 dataset has no missing values except for 676 entries in the site_id column.

In [48]:
# Finding out data types in Telcom dataset 3(Tel_3)
print(Tel_3.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5001 entries, 0 to 5000
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   PRODUCT          5001 non-null   object
 1   VALUE            5001 non-null   int64 
 2   DATE_TIME        5001 non-null   object
 3   CELL_ON_SITE     5001 non-null   object
 4   DW_A_NUMBER_INT  5001 non-null   object
 5   DW_B_NUMBER_INT  5001 non-null   object
 6   COUNTRY_A        5001 non-null   object
 7   COUNTRY_B        5001 non-null   object
 8   CELLID           5001 non-null   object
 9   SIET_ID          4396 non-null   object
dtypes: int64(1), object(9)
memory usage: 390.8+ KB
None


* Therew are no missing values in Tel_3 except for 605 entries in siet_id column

# Data Preparation and Data Cleaning

In [49]:
# Preparation of cgeo by running it.
cgeo

VILLES      STATUS LOCALISATION  ...          AREA     CELL_ID   SITE_CODE
0       ADJAME  In Service      ABIDJAN  ...           NaN  fbb2e75af3  0bb492ef17
1       ADJAME  In Service      ABIDJAN  ...           NaN  1502501d78  0bb492ef17
2       ADJAME  In Service      ABIDJAN  ...           NaN  0c2e182a0c  0bb492ef17
3      ASSINIE  In Service    INTERIEUR  ...           NaN  69bdfc69fd  8a1e6ce31f
4      ASSINIE  In Service    INTERIEUR  ...           NaN  7a1029d2ab  8a1e6ce31f
...        ...         ...          ...  ...           ...         ...         ...
3969  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  1cfc2d7ab1  f91ed8d2e9
3970  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  3ae1708153  f91ed8d2e9
3971  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  cb71f8a29c  61bee9c70c
3972  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  d4a6a0b645  61bee9c70c
3973  ZUENOULA  In Service    INTERIEUR  ...  CENTRE NORTH  f9f3e42589  61bee9c70c

[3974 rows x 11 columns]

In [50]:
# There are missing values in cell_geo(cgeo), so we replace the missing values in Status, Zonename and Area
cgeo["STATUS"] = cgeo['STATUS'].fillna("disconnected")
cgeo["ZONENAME"] = cgeo['ZONENAME'].fillna("Null")
cgeo["AREA"] = cgeo['AREA'].fillna("Null")

In [51]:
# Checking if cgeo information is updated.
cgeo.info()              # So it has updated the missing values.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3974 entries, 0 to 3973
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   VILLES        3974 non-null   object 
 1   STATUS        3974 non-null   object 
 2   LOCALISATION  3974 non-null   object 
 3   DECOUPZONE    3974 non-null   object 
 4   ZONENAME      3974 non-null   object 
 5   LONGITUDE     3974 non-null   float64
 6   LATITUDE      3974 non-null   float64
 7   REGION        3974 non-null   object 
 8   AREA          3974 non-null   object 
 9   CELL_ID       3974 non-null   object 
 10  SITE_CODE     3974 non-null   object 
dtypes: float64(2), object(9)
memory usage: 372.6+ KB


cleaning the telcom datasets

In [52]:
# Since we want to merge our Telcom datasets into one, we will need to synchronise the column names.
columns = ['Product','Value','Datetime','Cell_on_site','Number_a','Number_b','Country_A','Country_B','CELL_ID','Site_ID']
Tel_1.columns = columns
Tel_1.head()

Product  Value               Datetime  ... Country_B     CELL_ID     Site_ID
0   Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv

[5 rows x 10 columns]

In [53]:
Tel_2.columns = columns
Tel_2.head()

Product  Value               Datetime  ... Country_B     CELL_ID     Site_ID
0     sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1   Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2     sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3   Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4     sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO

[5 rows x 10 columns]

In [54]:

Tel_3.columns = columns
Tel_3.head()

Product  Value               Datetime  ... Country_B     CELL_ID     Site_ID
0   Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1     sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2   Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3     sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA

[5 rows x 10 columns]

In [55]:
 Telcom = pd.concat([Tel_1, Tel_2, Tel_3], ignore_index = 1)
 Telcom

Product  Value               Datetime  ... Country_B     CELL_ID     Site_ID
0       Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1         sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2         sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3       Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4       Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv
...       ...    ...                    ...  ...       ...         ...         ...
14998   Voice      0  2012-05-09 00:01:43.0  ...     19e2e  2aa04dcab7  f7d63a29e5
14999   Voice     62  2012-05-09 00:01:43.0  ...     19e2e  7f1f2f889f  b4ddb9e93c
15000   Voice     32  2012-05-09 00:01:43.0  ...     19e2e  a4731e780e  d68260ef50
15001   Voice     30  2012-05-09 00:01:43.0  ...     19e2e  98ed726da2  b4130b54ab
15002   Voice     60  2012-05-09 00:01:43.0  ...     19e2e  1c11a2fcbb  6d70dae094

[15003 rows x 10 columns]

In [56]:
# Dropping irrelevant columns
Clean_telcom = Telcom.drop(columns = ['Country_A' ,	'Country_B' ,'Cell_on_site'])
Clean_telcom.head()

Product  Value  ...     CELL_ID     Site_ID
0   Voice      0  ...  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  ...  /+cKZKIp41  /+cKZKIp41
2     sms     25  ...  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  ...  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  ...  051KhYDCpv  051KhYDCpv

[5 rows x 7 columns]

In [57]:
# Splitting date_time column into date and time
# create a new column and solit the date_time column into 2

Clean_telcom[['date', 'time']] = Clean_telcom.Datetime.str.split(expand=True)
Clean_telcom.head()

Product  Value               Datetime  ...     Site_ID        date        time
0   Voice      0  2012-05-06 23:04:37.0  ...  /TJNe+Mmtu  2012-05-06  23:04:37.0
1     sms      0  2012-05-06 23:05:25.0  ...  /+cKZKIp41  2012-05-06  23:05:25.0
2     sms     25  2012-05-06 23:05:41.0  ...  /u0FSD+ahi  2012-05-06  23:05:41.0
3   Voice     56  2012-05-06 23:05:42.0  ...  +cNeJzsTp3  2012-05-06  23:05:42.0
4   Voice     14  2012-05-06 23:06:29.0  ...  051KhYDCpv  2012-05-06  23:06:29.0

[5 rows x 9 columns]

In [58]:
# merge CDR and cgeo data.

cdr_cgeo = pd.merge(cgeo,Clean_telcom, how = 'inner',left_on = 'SITE_CODE', right_on = 'Site_ID')
cdr_cgeo.head(5)

VILLES      STATUS LOCALISATION  ...     Site_ID        date        time
0  ADJAME  In Service      ABIDJAN  ...  0bb492ef17  2012-05-07  00:00:58.0
1  ADJAME  In Service      ABIDJAN  ...  0bb492ef17  2012-05-07  00:00:58.0
2  ADJAME  In Service      ABIDJAN  ...  0bb492ef17  2012-05-07  00:00:58.0
3   ABOBO  In Service      ABIDJAN  ...  56306cc7ca  2012-05-07  00:00:18.0
4   ABOBO  In Service      ABIDJAN  ...  56306cc7ca  2012-05-07  00:00:30.0

[5 rows x 20 columns]

In [59]:
# Finding out number of most used product by grouping product and value entries.
Value = Clean_telcom.groupby(['Product'])['Value'].sum()
Value

Product
Voice    312593
data      17287
sms       53427
Name: Value, dtype: int64

* The most used product is voice at 312593 entries, data at 17287 entries, then sms at 53427 entries.

In [60]:
# Finding out number of most used cities for the three days
Most_used_cities = cdr_cgeo.groupby(['VILLES'])['Value'].sum().head()
Most_used_cities

VILLES
ABENGOUROU         8709
ABOBO            125921
ABOISSO            1650
ABOISSO COMOE       795
ABONGOUA              3
Name: Value, dtype: int64

* The most used Villes are Abengourou, Abobo, Aboisso, Aboiso comoe, Abongoua 